{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Igv-7g1DiMVj"
   },
   "source": [
    "# **Problem 4**\n",
    "\n",
    "***Note:*** You are only allowed to use pandas and numpy libraries if needed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dSb7tQXaiR-F"
   },
   "source": [
    "## **Part a):**\n",
    "\n",
    "Complete the following function to implement a dynamic programming algorithm that finds the optimal global alignment of two input DNA sequences.\n",
    "\n",
    "---\n",
    "**Inputs:**\n",
    "\n",
    "1- path_fasta_1 (string): path to \"Fasta\" format file of first sequence\n",
    "\n",
    "2- path_fasta_2 (string): path to \"Fasta\" format file of second sequence\n",
    "\n",
    "3- path_score_mat (string): path to substitution score matrix.\n",
    "\n",
    "4- gap_penalty (negative float): a enagtive real number. \n",
    "\n",
    "---\n",
    "\n",
    "**Outputs:** your function should return 3 values:\n",
    "\n",
    "1- The optimal global alignment score\n",
    "\n",
    "2- a string representing the alignment of first input sequence (any one, if more than one optimal solutions exist; use \"-\" for indels)\n",
    "\n",
    "3- a string representing the alignment of second input sequence (any one, if more than one optimal solutions exist; use \"-\" for indels)\n",
    "\n",
    "\n",
    "**Output example:** 5, \"ACCC--AC\", \"AC-CGGAC\"\n",
    "\n",
    "---\n",
    "\n",
    "**Note** since you will be implementing a dynamic programming alignment algorithm, we expect your program to be of **quadratic time** complexity (i.e., two sequences of length $n$ can be aligned in O($n^2$) time)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FRQ6x2qfCelv"
   },
   "source": [
    "## **Part b):**\n",
    "\n",
    "Run your alignment code created in the last part on the 10 provided pair of sequences (data_1 to data_10). For each alignment, use the provided substitution matrix (same for all 10 alignments) and gap penalty of -200. For each alignment $i$ ($i \\in \\{1..10\\}$) write the results in a text file named \"align_data_i.txt\" in the following format:\n",
    "\n",
    "\"The optimal alignment score between given sequences has score X\"\n",
    "\n",
    "\"aligned sequence 1\"\n",
    "\n",
    "\"aligned sequence 2\"\n",
    "\n",
    "---\n",
    "\n",
    "**Example:** for data_example, the output (align_data_example.txt) should be:\n",
    "\n",
    "The optimal alignment score between given sequences has score 3362.0\n",
    "\n",
    "CTCGTCCTAAAAAAAA-GGAACGCGTGACGCGCTAAAAG-CATTCGTGCC\n",
    "\n",
    "GTCGTCCTAAAAAAAAGGGAACGCGTGATGCGCTAAAAGACATCAGTGCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "id": "dgWURtxTh28A"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3362.0\n"
     ]
    }
   ],
   "source": [
    "# Part A\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def optimal_global_alignment(path_fasta_1, path_fasta_2, path_score_mat, gap_penalty = -200):\n",
    "    score = None\n",
    "    aligned_1 = \"\"\n",
    "    aligned_2 = \"\"\n",
    "    \n",
    "    # Read in data from path files\n",
    "    with open(path_fasta_1) as f1:\n",
    "        fasta_1 = f1.readlines()\n",
    "    \n",
    "    with open(path_fasta_2) as f2:\n",
    "        fasta_2 = f2.readlines()\n",
    "    \n",
    "    with open(path_score_mat) as sm:\n",
    "        score_mat = sm.read().split()\n",
    "    \n",
    "    # Assign inputted sequences as strings\n",
    "    sequence_1 = str(fasta_1[1]).strip()\n",
    "    sequence_2 = str(fasta_2[1]).strip()\n",
    "    \n",
    "    # Create a dictionary to determine match/mismatch score\n",
    "    scores_dict = {\n",
    "        \"AA\": score_mat[5],\n",
    "        \"AC\": score_mat[6],\n",
    "        \"AT\": score_mat[7],\n",
    "        \"AG\": score_mat[8],\n",
    "        \"CA\": score_mat[10],\n",
    "        \"CC\": score_mat[11],\n",
    "        \"CT\": score_mat[12],\n",
    "        \"CG\": score_mat[13],\n",
    "        \"TA\": score_mat[15],\n",
    "        \"TC\": score_mat[16],\n",
    "        \"TT\": score_mat[17],\n",
    "        \"TG\": score_mat[18],\n",
    "        \"GA\": score_mat[20],\n",
    "        \"GC\": score_mat[21],\n",
    "        \"GT\": score_mat[22],\n",
    "        \"GG\": score_mat[23],\n",
    "    }\n",
    "\n",
    "    # Create matrix for optimal scores\n",
    "    length_1 = len(sequence_1)\n",
    "    length_2 = len(sequence_2)\n",
    "        \n",
    "    optimal_scores = np.zeros([length_1 + 1, length_2 + 1], dtype = float)\n",
    "    \n",
    "    optimal_scores[0:(length_1 + 1), 0] = [i * gap_penalty for i in range(length_1 + 1)]\n",
    "    optimal_scores[0, 0:(length_2 + 1)] = [i * gap_penalty for i in range(length_2 + 1)]\n",
    "    \n",
    "    # Assign optimal score matrix values based on match, mismatch, and gaps\n",
    "    i = 1\n",
    "    while i <= length_1:\n",
    "        j = 1\n",
    "        while j <= length_2:\n",
    "            if sequence_1[i - 1] == sequence_2[j - 1]:\n",
    "                match = float(scores_dict[str(str(sequence_1[i - 1]) + str(sequence_2[j - 1]))])\n",
    "                optimal_scores[i][j] = optimal_scores[i - 1][j - 1] + match\n",
    "            else:\n",
    "                mismatch = float(scores_dict[str(str(sequence_1[i - 1]) + str(sequence_2[j - 1]))])\n",
    "                optimal_scores[i][j] = max(optimal_scores[i - 1][j - 1] + mismatch,\n",
    "                            optimal_scores[i - 1][j] + gap_penalty,\n",
    "                            optimal_scores[i][j - 1] + gap_penalty)\n",
    "            j += 1\n",
    "        i += 1\n",
    "    \n",
    "    # Assign score value\n",
    "    score = float(optimal_scores[length_1][length_2])\n",
    "\n",
    "    # Create array to store char of unicodes and corresponding index variables\n",
    "    sq1 = np.zeros(length_1 + length_2 + 1, dtype = int)\n",
    "    sq2 = np.zeros(length_1 + length_2 + 1, dtype = int)\n",
    "     \n",
    "    sq1_position = length_1 + length_2\n",
    "    sq2_position = length_1 + length_2\n",
    "    \n",
    "    i = length_1\n",
    "    j = length_2\n",
    "    \n",
    "    # Determine gap positions and put in correct location\n",
    "    while not (i == 0 or j == 0):\n",
    "        if sequence_1[i - 1] == sequence_2[j - 1]:       \n",
    "            sq1[sq1_position] = ord(sequence_1[i - 1])\n",
    "            sq2[sq2_position] = ord(sequence_2[j - 1])\n",
    "            sq1_position -= 1\n",
    "            sq2_position -= 1\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        \n",
    "        elif (optimal_scores[i - 1][j - 1] + float(scores_dict[str(str(sequence_1[i - 1]) + \n",
    "                                                                 str(sequence_2[j - 1]))])) == optimal_scores[i][j]:\n",
    "            sq1[sq1_position] = ord(sequence_1[i - 1])\n",
    "            sq2[sq2_position] = ord(sequence_2[j - 1])\n",
    "            sq1_position -= 1\n",
    "            sq2_position -= 1\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "         \n",
    "        elif (optimal_scores[i - 1][j] + gap_penalty) == optimal_scores[i][j]:\n",
    "            sq1[sq1_position] = ord(sequence_1[i - 1])\n",
    "            sq2[sq2_position] = ord('-')\n",
    "            sq1_position -= 1\n",
    "            sq2_position -= 1\n",
    "            i -= 1\n",
    "         \n",
    "        elif (optimal_scores[i][j - 1] + gap_penalty) == optimal_scores[i][j]:       \n",
    "            sq1[sq1_position] = ord('-')\n",
    "            sq2[sq2_position] = ord(sequence_2[j - 1])\n",
    "            sq1_position -= 1\n",
    "            sq2_position -= 1\n",
    "            j -= 1\n",
    "         \n",
    "    # Finish char arrays\n",
    "    while sq1_position > 0:\n",
    "        if i > 0:\n",
    "            i -= 1\n",
    "            sq1[sq1_position] = ord(sequence_1[i])\n",
    "            sq1_position -= 1\n",
    "        else:\n",
    "            sq1[sq1_position] = ord('-')\n",
    "            sq1_position -= 1\n",
    "     \n",
    "    while sq2_position > 0:\n",
    "        if j > 0:\n",
    "            j -= 1\n",
    "            sq2[sq2_position] = ord(sequence_2[j])\n",
    "            sq2_position -= 1\n",
    "        else:\n",
    "            sq2[sq2_position] = ord('-')\n",
    "            sq2_position -= 1\n",
    " \n",
    "    # Create aligned strings for output from char sequences\n",
    "    temp = 1\n",
    "    i = length_1 + length_2\n",
    "    while i >= 1:\n",
    "        if (chr(sq1[i]) == '-') and chr(sq2[i]) == '-':\n",
    "            temp = i + 1\n",
    "            break\n",
    "        i -= 1\n",
    "\n",
    "    i = temp\n",
    "    while i <= length_1 + length_2:\n",
    "        aligned_1 += chr(sq1[i])\n",
    "        i += 1\n",
    "         \n",
    "    i = temp\n",
    "    while i <= length_1 + length_2:\n",
    "        aligned_2 += chr(sq2[i])\n",
    "        i += 1\n",
    "    \n",
    "    # Print output\n",
    "#     print(f\"The optimal alignment score between given sequences has score = {score}\")\n",
    "#     print(aligned_1)\n",
    "#     print(aligned_2)\n",
    "    \n",
    "    return score, aligned_1, aligned_2\n",
    "\n",
    "# Test function using path files\n",
    "fasta_1 = r\"prob4_data\\data_example\\seq1.fasta\"\n",
    "fasta_2 = r\"prob4_data\\data_example\\seq2.fasta\"\n",
    "score_mat = r\"prob4_data\\data_example\\substitution_matrix.txt\"\n",
    "\n",
    "optimal_global_alignment(fasta_1, fasta_2, score_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Part B\n",
    "\n",
    "def test_optimal():\n",
    "    i = 1\n",
    "    for i in range(10):\n",
    "        fasta_1 = r\"prob4_data\\data_\" + str(i+1) + \"\\seq1.fasta\"\n",
    "        fasta_2 = r\"prob4_data\\data_\" + str(i+1) + \"\\seq2.fasta\"\n",
    "        score_mat = r\"prob4_data\\data_example\\substitution_matrix.txt\"\n",
    "        \n",
    "        f = open(\"align_data_\" + str(i+1) + \".txt\",\"w+\")\n",
    "        f.write(\"The optimal alignment score between given sequences has score \" + \n",
    "                str(optimal_global_alignment(fasta_1, fasta_2, score_mat)[0]) + \"\\n\" \n",
    "                + optimal_global_alignment(fasta_1, fasta_2, score_mat)[1] + \"\\n\"\n",
    "                + optimal_global_alignment(fasta_1, fasta_2, score_mat)[2])\n",
    "        f.close()\n",
    "\n",
    "test_optimal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "name": "HW2_problem4.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
